Трофимов Никита

WoeromProg opened the pull request:
@the-homeless-god 

1. Проект изменён под использование паттерна Result
2. Были добавлены тесты под новые изменения
3. Создан файл README.md с описанием состояния проекта
4. Были обработаны исключения

WoeromProg committed:
fp

the-homeless-god commented on TagsCloudContainer.csproj:
```diff
@@ -0,0 +1,28 @@
+    </ItemGroup>
+
+    <ItemGroup>
+      <Folder Include="Images\" />
```
Давай задокументируем это в README

the-homeless-god commented on TagsCloudContainer/App/ConsoleApp.cs:
```diff
@@ -0,0 +1,63 @@
+    private readonly FileReaderFactory _readerFactory;
+    private readonly Settings _settings;
+    private readonly WordProcessor _processor;
+    private readonly IDraw _draw;
```
Давай использовать свойства и не использовать _

the-homeless-god commented on TagsCloudContainer/App/ConsoleApp.cs:
```diff
@@ -0,0 +1,63 @@
+using TagsCloudContainer.WordProcessing;
+namespace TagsCloudContainer.App;
+
+public class ConsoleApp : IApp
```
Не следуешь SRP (single responsibility principle), у тебя точка входа занимается огромным количеством всего, исправить

the-homeless-god replied:
Все еще актуальный комментарий

the-homeless-god commented on TagsCloudContainer/App/ConsoleApp.cs:
```diff
@@ -0,0 +1,63 @@
+        IDraw draw)
+    {
+        _readerFactory = readerFactory;
+        _settings = settings ?? throw new ArgumentNullException(nameof(settings));
```
А обработки исключений нет

the-homeless-god commented on TagsCloudContainer/App/ConsoleApp.cs:
```diff
@@ -0,0 +1,63 @@
+    
+    public Result<None> Run()
+    {
+        var text = GetText(_settings.File).OnFail(Exit);
+        var boringText = GetText(_settings.BoringWordsFileName).OnFail(Exit);
+        var words = _processor.ProcessWords(text.Value, boringText.Value);
+        var bitmap = _draw.CreateImage(words).OnFail(Exit);
+        var projectDirectory = Directory.GetParent(Environment.CurrentDirectory).Parent.Parent.FullName;
+        var rnd = new Random();
+        bitmap.Value.Save(projectDirectory + "\\Images", $"Rectangles{rnd.Next(1, 1000)}", GetImageFormat(_settings.ImageFormat).GetValueOrThrow());
```
SRP

the-homeless-god replied:
и магические цифры и пути

the-homeless-god commented on TagsCloudContainer/App/ConsoleApp.cs:
```diff
@@ -0,0 +1,63 @@
+    {
+        return imageFormat.ToLower() switch
+        {
+            "png" =>  ImageFormat.Png.Ok(),
+            "jpeg" => ImageFormat.Jpeg.Ok(),
+            _ => Result.Fail<ImageFormat>($"Неверный формат изображения: {imageFormat}")
```
Можно использовать enum

the-homeless-god replied:
Все еще актуальный комментарий

the-homeless-god commented on TagsCloudContainer/App/ConsoleApp.cs:
```diff
@@ -0,0 +1,63 @@
+    
+    private Result<string> GetText(string filename)
+    {
+        if (!File.Exists(filename))
+            return Result.Fail<string>($"Файл не найден {filename}");
```
Использовать всегда { }

the-homeless-god commented on TagsCloudContainer/App/ConsoleApp.cs:
```diff
@@ -0,0 +1,63 @@
+        Console.WriteLine(text);
+        Environment.Exit(1);
+    }
+}
```
в конце файлов пустую строку добавить

the-homeless-god commented on TagsCloudContainer/App/Extensions/BitmapExtension.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace TagsCloudContainer.App.Extensions;
+
+public static class BitmapExtension
```
Почему статический?

the-homeless-god commented on TagsCloudContainer/App/Extensions/BitmapExtension.cs:
```diff
@@ -0,0 +1,12 @@
+{
+    public static void Save(this Bitmap bitmap, string path, string filename, ImageFormat format)
+    {
+        bitmap.Save($"{path}\\{filename}.{format}", format);
```
не хватает обработки исключений

the-homeless-god commented on TagsCloudContainer/Cloud/SpiralFunction.cs:
```diff
@@ -0,0 +1,26 @@
+            _step = step;
+        }
+
+        public Result<Point> GetNextPoint()
```
Нужны комментарии

the-homeless-god commented on TagsCloudContainer/Console/CommandLineOptions.cs:
```diff
@@ -0,0 +1,35 @@
+
+    [Option('b', "Boring", Required = false, HelpText = "Укажите путь до файла со скучными словами", Default = "../../../TextFiles/Boring.docx")]
+    public string PathToBoringWordsFile { get; set; }// =
+        //@"C:\Users\nikit\RiderProjects\di\TagsCloudContainer\TextFiles\Boring.txt";
```
?

the-homeless-god commented on TagsCloudContainer/DrawRectangle/RectangleDraw.cs:
```diff
@@ -0,0 +1,54 @@
+    }
+    public Result<Bitmap> CreateImage(List<Word> words)
+    {
+        var rectangles = _rectangleGenerator.GenerateRectangles(words);
+        if (!rectangles.IsSuccess)
+            return Result.Fail<Bitmap>(rectangles.Error);
+        var bitmap = CreateBitmap(rectangles.Value);
+        var shiftToBitmapCenter = new Size(bitmap.Width / 2, bitmap.Height / 2);
+        using var g = Graphics.FromImage(bitmap);
+        using var pen = new Pen(_settings.Color);
+        g.Clear(Color.Black);
+        var count = 0;
+        foreach (var word in words)
+        {
+            var rectangle = new Rectangle(
+                rectangles.Value[count].Location + shiftToBitmapCenter, 
+                rectangles.Value[count].Size);
+            if (rectangle.Left < 0
+                || rectangle.Right > bitmap.Width
+                || rectangle.Bottom > bitmap.Height
+                || rectangle.Top < 0)
+                return Result.Fail<Bitmap>($"Тэг не влез на изображение заданного размера: X: {_settings.CenterX} Y: {_settings.CenterY}");
+            using var font = new Font(_settings.FontName, word.Size);
+            g.DrawString(word.Value, font, pen.Brush, rectangle);
+            count++;
+        }
+
```
Разбить на подметоды

the-homeless-god commented on TagsCloudContainerTests/FileReaderTest.cs:
```diff
@@ -0,0 +1,33 @@
+    [TestFixture]
+    public class FileReaderTest
+    {
+        //Directory.GetParent(Environment.CurrentDirectory).Parent.Parent.FullName + "\\TextFiles\\" + fileName
```
?

WoeromProg committed:
Написал README

Удалил BitmapExtension.cs

В ConsoleApp.cs:
Переделал в свойства
Дал понятные названия переменным при сохранении изображения
Вынес часть логики в методы у метода Run
Убрал throw new ArgumentNullException у settings, посчитал не нужным

В RectangleDraw.cs:
Разбил на подметоды в CreateImage

Подправил пространства имен в файлах

Написал несколько тестов в WordProcessingTest
Добавил пару файлов в формате docx

Добавил {} у if

Убрал Result<Point> у GetNextPoint, посчитал лишним
Добавил пустую строку во всех файлах
Удалил комментарии, которые делал для себя в ходе написания

the-homeless-god commented on TagsCloudContainer.csproj:
```diff
@@ -0,0 +1,28 @@
+      <None Remove="Images\Rectangles863.Png" />
+      <None Remove="Images\Rectangles858.Png" />
+      <None Remove="Images\Rectangles305.Png" />
+    </ItemGroup>
```
Это точно тут нужно?

WoeromProg committed:
Рефакторинг: Вынес большую часть функционала из ConsoleApp.cs
Выделил основную часть функционала из ConsoleApp.cs в отдельные файлы ProcessWord.cs и PathImage.cs, которые размещены в новой директории AppExtensions.

the-homeless-god commented on TagsCloudContainer/App/AppExtensions/PathImage.cs:
```diff
@@ -0,0 +1,26 @@
+    {
+        return imageFormat.ToLower() switch
+        {
+            "png" =>  ImageFormat.Png.Ok(),
+            "jpeg" => ImageFormat.Jpeg.Ok(),
+            _ => Result.Fail<ImageFormat>($"Неверный формат изображения: {imageFormat}")
```
Давай будем использовать enum для ключей?

the-homeless-god commented on TagsCloudContainer/Cloud/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,63 @@
+        {
+            var nextPoint = spiralFunction.GetNextPoint();
+            rectangle = new Rectangle(nextPoint, sizeRectangle);
+            if(rectangle.IsIntersectOthersRectangles(_rectangles))
```
Куда-то делось форматирование кода + не хватает try-catch

WoeromProg replied:
Какую ошибку здесь можно отлавливать?

the-homeless-god commented on TagsCloudContainer/Cloud/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,63 @@
+
+    private Rectangle MoveRectangleToCenter(Rectangle rectangle)
+    {
+        var newRectangle = MoveRectangleAxis(rectangle, rectangle.GetCenter().X, _center.X, 
+            new Point(rectangle.GetCenter().X < _center.X ? 1 : -1, 0));
+        newRectangle = MoveRectangleAxis(newRectangle, newRectangle.GetCenter().Y, _center.Y, 
+            new Point(0, rectangle.GetCenter().Y < _center.Y ? 1 : -1));
+        return newRectangle;
```
Дублирование + магия + магические числа

the-homeless-god commented on TagsCloudContainer/Cloud/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,63 @@
+
+    private Rectangle MoveRectangleAxis(Rectangle newRectangle, int currentPosition, int desiredPosition, Point stepPoint)
+    {
+        while (newRectangle.IsIntersectOthersRectangles(_rectangles)  &&  desiredPosition != currentPosition)
+        {
+            currentPosition += currentPosition < desiredPosition ? 1 : -1;
+            newRectangle.Location = newRectangle.Location.DecreasingCoordinate(stepPoint);
+        }
+            
+        if (!newRectangle.IsIntersectOthersRectangles(_rectangles))
+        {
+            newRectangle.Location = newRectangle.Location.IncreasingCoordinate(stepPoint);
+        }
```
Опять дублирование + магия + отсутствие try-catch + сложно читать, а значит надо упрощать

the-homeless-god commented on TagsCloudContainer/Console/CommandLineOptions.cs:
```diff
@@ -0,0 +1,30 @@
+    [Option('i', "Input", Required = false, HelpText = "Укажите путь до файла", Default = "../../../TextFiles/Text.docx")]
+    public string PathToInputFile { get; set; }
+
+    [Option('b', "Boring", Required = false, HelpText = "Укажите путь до файла со скучными словами", Default = "../../../TextFiles/Boring.docx")]
```
Этот файл точно есть в репо?

WoeromProg replied:
![репо](https://github.com/kontur-courses/fp/assets/119159723/e301dd39-5269-4bd8-be2e-c59062c4828d)

the-homeless-god commented on TagsCloudContainer/DrawRectangle/RectangleDraw.cs:
```diff
@@ -0,0 +1,64 @@
+
+namespace TagsCloudContainer.DrawRectangle;
+
+public class RectangleDraw : IDraw
```
Отсутствие SOLID-а в целом, нужно дробить + оптимизировать

the-homeless-god commented on TagsCloudContainer/WordProcessing/Result.cs:
```diff
@@ -0,0 +1,83 @@
+    private None()
+    {
+    }
+}
```
Поплыло форматирование, плюс не стоит хранить по паре классов в одном файле

WoeromProg committed:
Рефакторинг: Разбил Result на части, вынес разные классы в разные файлы

WoeromProg committed:
feat: Добавил enum ImageFormat

WoeromProg committed:
Refactoring: Удалил ненужный ItemGroup

WoeromProg committed:
Refactor: изменил метод MoveRectangleAxis, убрал дублирование, добавил const ShiftPoint, удалил метод MoveRectangleToCenter т.к. стал не нужным.

WoeromProg committed:
Общие изменения, небольшой рефакторинг, просто коммит

WoeromProg committed:
Refactor: вынес CreateImage в отдельный класс DrawImage

WoeromProg committed:
Refactor: Переделал ConsoleApp под SRP, вынес часть логики в ImageWorker (Переименовал с ImageCreate).

WoeromProg committed:
Refactor: Вынес методы из MoveRectangleAxis

WoeromProg committed:
Refactor: Раздробил, часть вынес в расширение Rectangle, оптимизировал код.
Refactor: Изменил метод IsIntersecting в RectangleExtension.cs

WoeromProg committed:
Refactor: поправил форматирование

WoeromProg committed:
feat: Добавил OfAction, Then

WoeromProg committed:
feat: Новое изображение после изменений
